### ISSUE TITLE
unexpected oas3-valid-media-example error for schema property that overrides another's type

### ISSUE THREAD

**Issue by pedrodovale** [stoplightio/spectral#2110]
Hi!

I'm trying to define a common search shema shared amongst different Open API documents. One of these documents needs to extend the schema and change the type to array of one of its properties. Then it's defining an example for a request body that uses this schema. However, spectral is reporting an error, stating that the example should be the original type, rather than the array. 

Since other tools I'm using are able to cope with this, I assume that this is a bug in spectral. If that's not the case, can someone explain what I can do to improve the Open API document below? Thank you in advance.

I made a simplified document of what I'm trying to accomplish. Here's how to reproduce the bug:

Given this Open API document:

*openapi.yaml*
````yaml
openapi: 3.0.3
info:
  title: API
  version: 1.0.0
servers:
  - url: https://api.com/
paths:
  /:
    post:
      tags: []
      summary: Your GET endpoint
      description: Your GET endpoint
      operationId: get-endpoint
      requestBody:
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/ExtendedSearch"
            examples:
              example1:
                value:
                  criteria:
                    - field: 'name'
        required: true
      responses:
        "200":
          description: OK
          content:
            application/json:
              schema:
                properties:
                  error:
                    type: string
components:
  schemas:
    Search:
      type: object
      properties:
        criteria:
          type: object # criteria original type
    ExtendedSearch:
      allOf:
        - $ref: '#/components/schemas/Search'
        - type: object
          properties:
            criteria:
              type: array # override criteria original type
              items:
                $ref: '#/components/schemas/ExtendedSearchCriteria'
    ExtendedSearchCriteria:
      type: object
      properties:
        field:
          type: string
          enum: [ name, status, createdDate, lastModifiedDate ]

````

and this rule set:

*ruleset.yaml*
````yaml
extends:
  - spectral:oas
````

when I execute this command:

````
spectral lint --ruleset ruleset.yaml openapi.yaml
````

the output includes the following:

````
22:28    error  oas3-valid-media-example    "criteria" property type must be object    paths./.post.requestBody.content.application/json.examples.example1.value.criteria
````

contrary to what I expected:
- no errors reported

Some additional notes:
- I'm using Spectral v6.1.0
- Accordingly, Prism returns `must be object` for this this request:
````
curl -X POST http://127.0.0.1:4010/ --header "Content-Type: application/json" --data '{"criteria":[]}'
````

**Comment by P0lip** [comment]
Hey!
What you're trying to do here isn't quite valid.
One cannot use `allOf` for overriding.
The way `allOf` works can be summed up as "_each_ subchema needs to be valid".
There's no input that could match the following schema (simplified your example) 
```ts
allOf: [
 { type: 'array' }, // 1st subschema
 { type: 'object' } // 2nd subschema
]
```
If you provide an array, the validation against the second subschema will fail. If you provide an object, the first one will fail.
If your input is a primitive such as a string or a number, all subschemas will fail to validate successfully.

I'd say you should use smaller models that are compatible with each other. In such a case one may be able to use `allOf`, although one still must bear in mind the aforementioned behavior of `allOf`, so overall I'm not sure if using `allOf` for merging is something I'd recommend, as it's just extremely tricky to maintain as your schema definitions grow over time.
It's basically just a hack.

I hope that helps. 
Let me know down here if you have anything further questions.

**Comment by pedrodovale** [comment]
Hi @P0lip,

Yes, I understand it's something you shouldn't need to do. In fact, we went in the direction you suggested:

> use smaller models that are compatible with each other

But since I didn't find anything in the spec preventing me to this and that there are other tools that still process the OpenAPI document the way I expect it to (e.g. SwaggerUI), I assumed it's valid approach. Can you provide me with reference documentation on why it isn't?

Thank you.

**Comment by P0lip** [comment]
Hey!

> But since I didn't find anything in the spec preventing me to this and that there are other tools that still process the 
OpenAPI document the way I expect it to (e.g. SwaggerUI), I assumed it's valid approach. 

I don't know how SwaggerUI handles that case, but if they give an impression allOf works like an override, it sounds like a bug to me. They shouldn't process that in such a way if they do so.
Overall though certain tools or libraries allow some soft mistakes that aren't syntax errors.

>  Can you provide me with reference documentation on why it isn't?

https://json-schema.org/draft/2020-12/json-schema-core.html#rfc.section.10.2.1.1

> This keyword's value MUST be a non-empty array. Each item of the array MUST be a valid JSON Schema.
> **An instance validates successfully against this keyword if it validates successfully against all schemas defined by this keyword's value.**

**Comment by pedrodovale** [comment]
Hi @P0lip,

Yes, I understand it's something you shouldn't need to do. In fact, we went in the direction you suggested:

> use smaller models that are compatible with each other

But since I didn't find anything in the spec preventing me to this and that there are other tools that still process the OpenAPI document the way I expect it to (e.g. SwaggerUI), I assumed it's valid approach. Can you provide me with reference documentation on why it isn't?

Thank you.
